mne/viz/_figure.py

```diff
@@ -0,0 +1,1854 @@
+        labels = [label.get_text() for label in buttons.labels]
+        inds = np.in1d(parent.mne.ch_names, chs)
+        parent.mne.ch_selections['Custom'] = inds.nonzero()[0]
+        buttons.set_active(labels.index('Custom'))
+
+
+class MNEBrowseFigure(MNEFigure):
+    """Interactive figure with scrollbars, for data browsing."""
+
+    def __init__(self, inst, figsize, xlabel='Time (s)', **kwargs):
+        from matplotlib import rcParams
+        from matplotlib.colors import to_rgba_array
+        from matplotlib.patches import Rectangle
+        from mpl_toolkits.axes_grid1.axes_size import Fixed
+        from mpl_toolkits.axes_grid1.axes_divider import make_axes_locatable
+        from .. import BaseEpochs
+        from ..io import BaseRaw
+        from ..preprocessing import ICA
+
+        fgcolor = rcParams['axes.edgecolor']
+        bgcolor = rcParams['axes.facecolor']
+
+        super().__init__(figsize=figsize, inst=inst, **kwargs)
+
+        # what kind of data are we dealing with?
+        if isinstance(inst, BaseRaw):
+            self.mne.instance_type = 'raw'
+        elif isinstance(inst, BaseEpochs):
+            self.mne.instance_type = 'epochs'
+        elif isinstance(inst, ICA):
+            self.mne.instance_type = 'ica'
+        else:
+            raise TypeError('Expected an instance of Raw, Epochs, or ICA, '
+                            f'got {type(inst)}.')
+
+        # additional params for browse figures
+        self.mne.ch_start = 0
+        self.mne.projector = None
+        self.mne.projs_active = np.array([p['active'] for p in self.mne.projs])
+        self.mne.whitened_ch_names = list()
+        self.mne.use_noise_cov = self.mne.noise_cov is not None
+        self.mne.zorder = dict(ann=0, events=1, bads=2, data=3, mag=4, grad=5,
+                               scalebar=6, vline=7)
+        # annotations
+        self.mne.annotations = list()
+        self.mne.hscroll_annotations = list()
+        self.mne.annotation_segments = list()
+        self.mne.annotation_texts = list()
+        self.mne.new_annotation_labels = list()
+        self.mne.annotation_segment_colors = dict()
+        self.mne.annotation_hover_line = None
+        self.mne.draggable_annotations = False
+        # lines
+        self.mne.event_lines = None
+        self.mne.event_texts = list()
+        self.mne.vline_visible = False
+        # scalings
+        self.mne.scale_factor = 0.5 if self.mne.butterfly else 1.
+        self.mne.scalebars = dict()
+        self.mne.scalebar_texts = dict()
+        # ancillary child figures
+        self.mne.child_figs = list()
+        self.mne.fig_help = None
+        self.mne.fig_proj = None
+        self.mne.fig_selection = None
+        self.mne.fig_annotation = None
+
+        # MAIN AXES: default sizes (inches)
+        l_margin = 1.
+        r_margin = 0.1
+        b_margin = 0.45
+        t_margin = 0.25
+        scroll_width = 0.25
+        hscroll_dist = 0.25
+        vscroll_dist = 0.1
```

Comment:
matplotlib 2.2 has `constrained_layout=True`. Maybe add a comment that we might be able to simplify this stuff at some point by bumping our min and using it?